fix flaky test in 17

MyEnthusiastic opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*(For example: This pull request makes task deployment go through the blob server, rather than through RPC. That way we avoid re-transferring them on each deployment (during recovery).)*


## Brief change log

*(for example:)*
  - *The TaskInfo is stored in the blob store on job creation time as a persistent artifact*
  - *Deployments RPC transmits only the blob storage reference*
  - *TaskManagers retrieve the TaskInfo from the blob cache*


## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (100MB)*
  - *Extended integration test for recovery after master (JobManager) failure*
  - *Added test that validates that TaskInfo is transferred only once across recoveries*
  - *Manually verified the change by running a 4 node cluster with 2 JobManagers and 4 TaskManagers, a stateful streaming program, and killing one JobManager and two TaskManagers during the execution, verifying that recovery happens correctly.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know)
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

zzjas commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/OperatorStateBackendTest.java:
```diff
@@ -649,20 +649,23 @@ void testSnapshotRestoreSync() throws Exception {
             bIt = broadcastState2.iterator();
             assertThat(bIt).hasNext();
-            entry = bIt.next();
+            Map.Entry<Serializable, Serializable> entry = bIt.next();
```
The original test reuses this variable. Maybe you could also reuse it with the one at Line 654.

MyEnthusiastic replied:
Resolved : Already made that resuable

Suraj-Vashista-BK commented:
> ## What is the purpose of the change
> _Fix flaky test_
> 
> ## Brief change log
> * _broadcastState2 is kind of **map** (although I'm not able to find implementation, only python version of implementation in that repository, doubt it is transferred to java as somewhere. I infer that mainly from the put operation and flaky test result )_
> * _we put two elements into it, but when test, we assume the first entry we get is the <1, 2> and the second entry is <2, 5> which is not necessary right._
> 
> ## Verifying this change
> Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing
> 
> _(Please pick either of the following options)_
> 
> * _This change is to fix the test, it changes the test itself_
> 
> ## Does this pull request potentially affect one of the following parts:
> * Dependencies (does it add or upgrade a dependency): no
> * The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
> * The serializers: no
> * The runtime per-record code paths (performance sensitive): no
> * Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
> * The S3 file system connector: no
> 
> ## Documentation
> * Does this pull request introduce a new feature? no
> * If yes, how is the feature documented? not applicable
> 
> Here is the script for running the non-dex
> 
> ```
> mvn -pl flink-runtime -DnondexRuns=10 edu.illinois:nondex-maven-plugin:2.1.7-SNAPSHOT:nondex -Dtest=org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreSync```
> ```

The fix looks okay. One suggestion would be to link you account to the commit. If you check your commit, it does not link to your git account. When you raise a real PR , if your commit is not linked to your git, you wont appear as a contributor.

yashdeep97 commented:
Nice work on the code changes!
Minor comment on the PR description - can you add a short description of the problem and the cause of flakiness in the tests and how the code change addresses the flakiness.

MyEnthusiastic commented:
> > ## What is the purpose of the change
> > _Fix flaky test_
> > ## Brief change log
> > 
> > * _broadcastState2 is kind of **map** (although I'm not able to find implementation, only python version of implementation in that repository, doubt it is transferred to java as somewhere. I infer that mainly from the put operation and flaky test result )_
> > * _we put two elements into it, but when test, we assume the first entry we get is the <1, 2> and the second entry is <2, 5> which is not necessary right._
> > 
> > ## Verifying this change
> > Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing
> > _(Please pick either of the following options)_
> > 
> > * _This change is to fix the test, it changes the test itself_
> > 
> > ## Does this pull request potentially affect one of the following parts:
> > 
> > * Dependencies (does it add or upgrade a dependency): no
> > * The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
> > * The serializers: no
> > * The runtime per-record code paths (performance sensitive): no
> > * Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
> > * The S3 file system connector: no
> > 
> > ## Documentation
> > 
> > * Does this pull request introduce a new feature? no
> > * If yes, how is the feature documented? not applicable
> > 
> > Here is the script for running the non-dex
> > ```
> > mvn -pl flink-runtime -DnondexRuns=10 edu.illinois:nondex-maven-plugin:2.1.7-SNAPSHOT:nondex -Dtest=org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreSync```
> > ```
> 
> The fix looks okay. One suggestion would be to link you account to the commit. If you check your commit, it does not link to your git account. When you raise a real PR , if your commit is not linked to your git, you wont appear as a contributor.



> > ## What is the purpose of the change
> > _Fix flaky test_
> > ## Brief change log
> > 
> > * _broadcastState2 is kind of **map** (although I'm not able to find implementation, only python version of implementation in that repository, doubt it is transferred to java as somewhere. I infer that mainly from the put operation and flaky test result )_
> > * _we put two elements into it, but when test, we assume the first entry we get is the <1, 2> and the second entry is <2, 5> which is not necessary right._
> > 
> > ## Verifying this change
> > Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing
> > _(Please pick either of the following options)_
> > 
> > * _This change is to fix the test, it changes the test itself_
> > 
> > ## Does this pull request potentially affect one of the following parts:
> > 
> > * Dependencies (does it add or upgrade a dependency): no
> > * The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
> > * The serializers: no
> > * The runtime per-record code paths (performance sensitive): no
> > * Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
> > * The S3 file system connector: no
> > 
> > ## Documentation
> > 
> > * Does this pull request introduce a new feature? no
> > * If yes, how is the feature documented? not applicable
> > 
> > Here is the script for running the non-dex
> > ```
> > mvn -pl flink-runtime -DnondexRuns=10 edu.illinois:nondex-maven-plugin:2.1.7-SNAPSHOT:nondex -Dtest=org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreSync```
> > ```
> 
> The fix looks okay. One suggestion would be to link you account to the commit. If you check your commit, it does not link to your git account. When you raise a real PR , if your commit is not linked to your git, you wont appear as a contributor.

Resolved! Thanks for reminding me

MyEnthusiastic commented:
> Nice work on the code changes! Minor comment on the PR description - can you add a short description of the problem and the cause of flakiness in the tests and how the code change addresses the flakiness.

Resolved! make it more clear, included what was the case before change and what change can bring

zzjas commented:
Just noticed, if you haven't opened a real PR yet, you would want a more descriptive title as well.

MyEnthusiastic commented:
This PR has already been opened : https://github.com/apache/flink/pull/23464 


And the logic should be same for the test:

- org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreSync
- org.apache.flink.runtime.state.OperatorStateBackendTest#testSnapshotRestoreAsync
